### PR TITLE
nixos/make-disk-image: fix contents dir paths

### PR DESCRIPTION
**Pull request by leon-barrett**
###### Description of changes
`make-disk-image` is a tool for creating VM images. It takes an argument `contents` that allows one to specify files and directories that should be copied into the VM image. However, directories do not end up at the specified target, but instead at a subdirectory of the target, with a nix-store-like path, e.g. `/target/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-source`. See issue https://github.com/NixOS/nixpkgs/issues/226203 .

This change adds a test for make-disk-image's contents directory handling and adds a fix (appending `/` to rsync input directory names).

This closes issue https://github.com/NixOS/nixpkgs/issues/226203 .

###### Things done
I added some checks in `nixosTests.image-contents` and ran it on an x86_64 machine. I also ran `nixosTests.qemu-vm-restrictnetwork`.

I did not run these relevant-looking tests, however, because they had issues on master:
* `nixosTests.ec2-config` (not run because it logged about something broken on master)
* `nixosTests.ec2-nixops` (not run because it failed on master)
* `nixosTests.sourcehut` (not tested because it did not complete in 10 minutes on master)
* `nixosTests.systemd-repart` (not tested because it failed on master)

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [X] [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)), see specific tests listed above
  - [ ] and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - [ ] or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - [ ] made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [X] Added a release notes entry for this library function fix. (I'm not sure if this is useful.)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/make-disk-image.nix [hunk 1]
```diff
@@ -402,11 +402,13 @@ let format' = format; in let
...
       else
         mkdir -p $root/$(dirname $target)
-        if ! [ -e $root/$target ]; then
-          rsync $rsync_flags $source $root/$target
-        else
+        if [ -e $root/$target ]; then
           echo "duplicate entry $target -> $source"
           exit 1
+        elif [ -d $source ]; then
+          rsync $rsync_flags $source/ $root/$target
```
Comment (RaitoBezarius) [inline nixos/lib/make-disk-image.nix:409]: Please add a comment to explain the rsync behavioral differences (everyone always gets tripped on this.)

File: nixos/tests/common/ec2.nix [hunk 2]
```diff
@@ -59,7 +62,11 @@ with pkgs.lib;
         )
 
         machine = create_machine({"startCommand": start_command})
-      '' + script;
+        try:
+      '' + indentLines(script) + ''
```
Comment (RaitoBezarius) [inline nixos/tests/common/ec2.nix:66]:
```suggestion
      '' + indentLines script + ''
```

File: nixos/tests/common/ec2.nix [hunk 3]
```diff
@@ -36,6 +37,8 @@ with pkgs.lib;
                 "create",
                 "-f",
                 "qcow2",
+                "-F",
+                "qcow2",
```
Comment (RaitoBezarius) [inline nixos/tests/common/ec2.nix:41]: Modification of this file should go in a separate commit to keep clean history

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Minor changes are necessary, but LGTM other than that, thank you!

File: nixos/tests/common/ec2.nix (hunk 3 above)
Comment (leon-barrett) [reply nixos/tests/common/ec2.nix:41]: I moved these changes to another commit in the same PR. Should I move them to another PR?

File: nixos/lib/make-disk-image.nix (hunk 1 above)
Comment (leon-barrett) [reply nixos/lib/make-disk-image.nix:412]: Done.

File: nixos/tests/common/ec2.nix (hunk 3 above)
Comment (RaitoBezarius) [reply nixos/tests/common/ec2.nix:41]: No, but the commit should be formatted using a path name: `nixos/tests/ec2` is better

**Review by RaitoBezarius — APPROVED** [review]
Minor change on the commit formatting for EC2 and we are good to go!

File: nixos/tests/common/ec2.nix (hunk 3 above)
Comment (leon-barrett) [reply nixos/tests/common/ec2.nix:41]: Okay, thanks, I wasn't sure quite how to format it.

**Review by RaitoBezarius — APPROVED** [review]

**Comment by leon-barrett** [comment]
Thanks!
